Add world builder integration to CPO initial grains.

MFraters opened the pull request:
This pull request enables the use of the world builder to setup initial grain-sizes and initial grain orientations for different regions in the model.

I do still need to add tests. @Wang-yijun Can you also take a look at this to see if it works for you?

Wang-yijun commented:
I've tested using initial textures from GWB on this branch. It works for me!

MFraters commented:
@Wang-yijun Thanks for testing!

This pull request is now ready for review

gassmoeller commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -159,12 +162,33 @@ namespace aspect
+                  {
+                    sum_volume_fractions += wb_grains.sizes[grain_i];
+                    volume_fractions_grains[mineral_i][grain_i] = wb_grains.sizes[grain_i];
+                    // we are receiving a array<array<double,3>,3> which needs to be unrolled in the correct way
+                    // for a tensor<2,3> it just loops first over the second index and than the first index
```
This comment needs some punctuation to clarify the meaning

gassmoeller commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -159,12 +162,33 @@ namespace aspect
+                        for (unsigned int component_j = 0; component_j < 3 ; ++component_j)
+                          {
+                            Assert(!std::isnan(wb_grains.rotation_matrices[grain_i][component_i][component_j]), ExcMessage("Error: not a number."));
+                            rotation_matrices_grains[mineral_i][grain_i][component_i][component_j] = wb_grains.rotation_matrices[grain_i][component_i][component_j];
```
Can you add a comment here that explains why you can assign the `component_i` of `wb_grains` to all `component_j` of `rotation_matrices_grains`?

MFraters replied:
I don't understand what you mean here. component i,j is assigned to component i,j

gassmoeller commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -159,12 +162,33 @@ namespace aspect
+
+                  }
+
+                AssertThrow(sum_volume_fractions != 0, ExcMessage("Sum of volumes is equal to zero, which is not supporsed to happen. "
```
```suggestion
                AssertThrow(sum_volume_fractions != 0, ExcMessage("Sum of volumes is equal to zero, which is not supposed to happen. "
```

gassmoeller commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -1156,9 +1180,21 @@ namespace aspect
-                AssertThrow(model_name == "Uniform grains and random uniform rotations",
-                            ExcMessage("No model named " + model_name + "for CPO particle property initialization. "
-                                       + "Only the model \"Uniform grains and random uniform rotations\" is available."));
+                if (model_name  == "Uniform grains and random uniform rotations")
```
```suggestion
                if (model_name == "Uniform grains and random uniform rotations")
```

gassmoeller commented on tests/world_builder_select_grains.prm:
```diff
@@ -0,0 +1,114 @@
+# Enable if: ASPECT_WITH_WORLD_BUILDER
+# This test is a copy of world_builder_simple.prm and checks that we can select
+# a subset of compositions for which to evaluate the world builder.
```
I think this comment needs an update.

gassmoeller replied:
Also, if this is just an update, can you `include` the other test .prm file?

gassmoeller commented on tests/world_builder_select_grains/particles/particles-00000.0000.vtu:
could you change the output format to `gnuplot` to simplify the test comparison?

gassmoeller reviewed (commented):
Very nice, a few comments. 
Does this PR work cleanly with the world builder version included in ASPECT, or do you need a development version? If you need a development version we need a check for WORLD_BUILDER_VERSION in this plugin.

MFraters commented:
@gassmoeller I addressed your comments, but I didn't understand what you meant with one of them. This pull request works with the world builder which is part of aspect. With the changes @Wang-yijun made, better tests can be made, but that will have to wait until the new wold builder version is released and becomes part of aspect.

gassmoeller reviewed (commented):
Very nice, thanks. I found a few more minor points.
Could you also add a changelog entry? As far as I see this is a new feature that was added.

MFraters commented:
@gassmoeller Thanks for the review. I have addressed your comments

gassmoeller commented on include/aspect/particle/property/crystal_preferred_orientation.h:
```diff
@@ -86,6 +86,17 @@ namespace aspect
       };
 
+      /**
+       * @brief An enum used to deterime how the initial grain sizes and orientation are set for all particles
```
```suggestion
       * @brief An enum used to determine how the initial grain sizes and orientations are set for all particles
```

gassmoeller reviewed (approved):
One typo left to fix, afterwards feel free to merge yourself.
